subprojects/core/src/main/java/org/gradle/process/internal/DefaultMemoryResourceManager.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright 2016 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.process.internal;
+
+import com.google.common.base.Preconditions;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+import org.gradle.process.internal.health.memory.MemoryInfo;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+
+public class DefaultMemoryResourceManager implements MemoryResourceManager {
+
+    private static final Logger LOGGER = Logging.getLogger(MemoryResourceManager.class);
+    private static final long MIN_THRESHOLD_BYTES = 384 * 1024 * 1024; // 384M
+    private static final int MONITOR_CHECK_INITIAL_DELAY_SECONDS = 5;
+    private static final int MONITOR_CHECK_INTERVAL_SECONDS = 5;
+
+    private final MemoryInfo memoryInfo;
+    private final long memoryThresholdInBytes;
+    private final Object lock = new Object();
+    private final List<MemoryResourceHolder> holders = new ArrayList<MemoryResourceHolder>();
+
+    public DefaultMemoryResourceManager(ScheduledExecutorService scheduledExecutorService, MemoryInfo memoryInfo, double minFreeMemoryPercentage) {
+        Preconditions.checkArgument(minFreeMemoryPercentage >= 0, "Free memory percentage must be >= 0");
+        Preconditions.checkArgument(minFreeMemoryPercentage <= 1, "Free memory percentage must be <= 1");
+        this.memoryInfo = Preconditions.checkNotNull(memoryInfo);
+        this.memoryThresholdInBytes = Math.max(MIN_THRESHOLD_BYTES, (long) (memoryInfo.getTotalPhysicalMemory() * minFreeMemoryPercentage));
+        scheduleMonitoringThread(scheduledExecutorService);
+    }
+
+    private void scheduleMonitoringThread(ScheduledExecutorService scheduledExecutorService) {
+        scheduledExecutorService.scheduleAtFixedRate(new FreeMemoryCheck(), MONITOR_CHECK_INITIAL_DELAY_SECONDS, MONITOR_CHECK_INTERVAL_SECONDS, TimeUnit.SECONDS);
```

Comment:
I think we should reuse the same periodic probe that the daemon expiry uses (not the same expiry algorithm though). So, for example, we would have a single scheduled job that calls `MemoryInfo.getFreePhysicalMemory()` and broadcasts an event with that value in it. Both the daemon expiry and the worker expiry can receive this event and act on it. We might also use the same memory threshold in both places to mean 'system is getting low on memory' so that the memory probe would fire an event on low memory instead. And another event to mean 'system memory is ok again'.

There's other ways to structure this, but the basic idea is to have the daemon expiry and the worker expiry work together in some way and to have the same concept of 'low system memory'. We should also prefer killing off worker processes on low memory, if we can, rather than have the daemon exit at the end of the build. We should certainly change the daemon expiry (maybe not for this story though) to 'unexpire' the daemon if system memory drops below the threshold by the end of the build.